scripts/eurostat/regional_statistics_by_nuts/population_density/PopulationDensity_preprocess_gen_tmcf.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pandas as pd
+import io
+import csv
+
+# data_url = "https://ec.europa.eu/eurostat/estat-navtree-portlet-prod/BulkDownloadListing?file=data/demo_r_d3area.tsv.gz"
+
+_SOURCE_TSV = "./demo_r_d3dens.tsv"
+_SOURCE_CSV_LONG = "./demo_r_d3dens.csv"
+_CLEANED_CSV = "./PopulationDensity_Eurostat_NUTS3.csv"
+_TMCF = "./PopulationDensity_Eurostat_NUTS3.tmcf"
+
+_OUTPUT_COLUMNS = [
+    'Date',
+    'GeoId',
+    'Count_Person_PerArea',
+]
+
+
+def translate_wide_to_long(source_tsv, source_csv_long):
+    df = pd.read_csv(source_tsv, delimiter='\t')
+    assert df.head
+
+    header = list(df.columns.values)
+    years = header[1:]
+
+    # Pandas.melt() unpivots a DataFrame from wide format to long format.
+    df = pd.melt(df,
+                 id_vars=header[0],
+                 value_vars=years,
+                 var_name='time',
+                 value_name='value')
+
+    # Separate geo and unit columns.
+    new = df[header[0]].str.split(",", n=1, expand=True)
+    df['geo'] = new[1]
+    df['unit'] = new[0]
+    df.drop(columns=[header[0]], inplace=True)
+
+    # Remove empty rows, clean values to have all digits.
+    df = df[df.value.str.contains('[0-9]')]
+    possible_flags = [' ', ':', 'b', 'e']
+    for flag in possible_flags:
+        df['value'] = df['value'].str.replace(flag, '')
+
+    df['value'] = pd.to_numeric(df['value'])
+    df.to_csv(source_csv_long, index=False)
+
+
+def preprocess(cleaned_csv, source_csv_long, output_cols):
+    with open(cleaned_csv, 'w', newline='') as f_out:
+        writer = csv.DictWriter(f_out,
+                                fieldnames=output_cols,
+                                lineterminator='\n')
+        with open(source_csv_long) as response:
+            reader = csv.DictReader(response)
```

Comment:
Let's pass the data frame into the function rather than reading it from a file, simplifying the code.